src/Google.Datastore.V1Beta3/ArrayValuePartial.cs

```diff
@@ -0,0 +1,82 @@
+        public static implicit operator ArrayValue(DateTimeOffset[] values) => ToArrayValue(values, x => x);
+        public static implicit operator ArrayValue(DateTimeOffset?[] values) => ToArrayValue(values, x => x);
+
+        private static ArrayValue ToArrayValue<T>(T[] values, Func<T, Value> converter) =>
+            values == null ? null : new ArrayValue { Values = { values.Select(x => converter(x) ?? Value.ForNull()) } };
+
+        public static explicit operator string[](ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (string) x);
+        public static explicit operator long[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (long)x);
+        public static explicit operator long?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (long?)x);
+        public static explicit operator Key[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (Key)x);
+        public static explicit operator double[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (double)x);
+        public static explicit operator double?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (double?)x);
+        public static explicit operator bool[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (bool)x);
+        public static explicit operator bool?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (bool?)x);
+        public static explicit operator ArrayValue[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (ArrayValue)x);
+        public static explicit operator ByteString[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (ByteString)x);
+        public static explicit operator byte[][] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (byte[])x);
+        public static explicit operator Entity[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (Entity)x);
+        public static explicit operator LatLng[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (LatLng)x);
+        public static explicit operator Timestamp[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (Timestamp)x);
+        public static explicit operator DateTime[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTime)x);
+        public static explicit operator DateTime?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTime?)x);
+        public static explicit operator DateTimeOffset[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTimeOffset)x);
+        public static explicit operator DateTimeOffset?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTimeOffset?)x);
+
+        private static T[] FromArrayValue<T>(ArrayValue arrayValue, Func<Value, T> converter) =>
+            arrayValue == null ? null : arrayValue.Values.Select(x => converter(x)).ToArray();
+
+        public IEnumerator<Value> GetEnumerator() => Values.Select(x => x.OrNull()).GetEnumerator();
```

Comment:
The aim was to try to be more "idiomatic C#" where null is represented by a null reference, not a Value with a particular property. In particular, I think it's useful to be able to call Add(null) and end up with a Value... so this is sort of the opposite of that.

I may punt on this aspect of the change for now, as I strongly suspect it's the most potentially confusing one.

But yes, +1 to Value.IsNull. Will add that.

(And yes, the NullValue property is a little odd... this is just the common pattern for representing a oneof case in protos where the only information required is which case is being set.)